### ISSUE TITLE
Add MG test for false fallback activation

### ISSUE THREAD

**Issue by GitGab19** [stratum-mining/stratum#923]
This is more a reminder for me than an issue, but I preferred to open an issue to better track it.
We need to add a MG test for specific scenario described in issue #901, addressing suggestions given under PR #904, that has been merged yesterday.

**Comment by Shourya742** [comment]
taking this up

**Comment by GitGab19** [comment]
Hey @Shourya742 I had a look at suggestions mentioned in this issue description. Here's the comments by @Fi3 on the old PR which fixed the previous issue:
- https://github.com/stratum-mining/stratum/pull/904#discussion_r1609437598
- https://github.com/stratum-mining/stratum/pull/904#discussion_r1609452850

Don't know if @Fi3 still thinks they are relevant changes, and if you're willing to address them btw

**Comment by Shourya742** [comment]
> Hey @Shourya742 I had a look at suggestions mentioned in this issue description. Here's the comments by @Fi3 on the old PR which fixed the previous issue:
> 
> * [Better management of LastDeclareJobs - no more wrong fallback-system activations #904 (comment)](https://github.com/stratum-mining/stratum/pull/904#discussion_r1609437598)
> * [Better management of LastDeclareJobs - no more wrong fallback-system activations #904 (comment)](https://github.com/stratum-mining/stratum/pull/904#discussion_r1609452850)
> 
> Don't know if @Fi3 still thinks they are relevant changes, and if you're willing to address them btw

I managed to move past the error where plebhash left off, but now I've hit a few new issues. It's about the lastDeclareJob. I was looking over the JDC, and it definitely makes sense to add these tweaks as we go. I'll put together a PR for this too.


<details>
<summary>Error Message</summary>
![Screenshot from 2024-06-28 04-36-28](https://github.com/stratum-mining/stratum/assets/84662239/9f03a7a3-eb40-4104-851f-72637eecae0c)
</details>

**Comment by GitGab19** [comment]
I think that your screenshot has not been uploaded correctly. I get this:
```
Error Message
![Screenshot from 2024-06-28 04-36-28](https://github.com/stratum-mining/stratum/assets/84662239/9f03a7a3-eb40-4104-851f-72637eecae0c) 
```

**Comment by Shourya742** [comment]
```
STD OUT:     ),
STD OUT: )
STD OUT: Working on result 1/1: MatchMessageType: 87 (0x57)
STD OUT: 2024-06-28T08:35:02.218217Z DEBUG roles_logic_sv2::handlers::job_declaration: Received AllocateMiningJobTokenSuccess with id: 1
STD OUT: 2024-06-28T08:35:02.262600Z  INFO roles_logic_sv2::handlers::job_declaration: Received DeclareMiningJobSuccess with id 0
STD OUT: 2024-06-28T08:35:02.262618Z DEBUG roles_logic_sv2::handlers::job_declaration: DeclareMiningJobSuccess: DeclareMiningJobSuccess { request_id: 0, new_mining_job_token: Ref([0, 0, 0, 0, 0]) }
STD OUT: 2024-06-28T08:35:02.262644Z  INFO jd_client::lib::job_declarator: Wth JobDeclarator { receiver: Receiver { .. }, sender: Sender { .. }, allocated_tokens: [AllocateMiningJobTokenSuccess { request_id: 0, mining_job_token: Owned([0, 0, 0, 0, 0]), coinbase_output_max_additional_size: 0, coinbase_output: Owned([]), async_mining_allowed: false }, AllocateMiningJobTokenSuccess { request_id: 1, mining_job_token: Owned([0, 0, 0, 0, 1]), coinbase_output_max_additional_size: 0, coinbase_output: Owned([]), async_mining_allowed: false }], req_ids: Id { state: 0 }, min_extranonce_size: 8, last_declare_mining_jobs_sent: {}, last_set_new_prev_hash: None, set_new_prev_hash_counter: 0, future_jobs: {}, up: Mutex(Mutex { data: Upstream { channel_id: None, tx_status: Upstream(Sender { .. }), min_extranonce_size: 0, upstream_extranonce1_size: 16, pool_signature: "Stratum v2 SRI Pool", receiver: Receiver { .. }, sender: Sender { .. }, downstream: None, task_collector: Mutex(Mutex { data: [AbortHandle { id: Id(12) }, AbortHandle { id: Id(15) }], poisoned: false, .. }), pool_chaneger_trigger: Mutex(Mutex { data: PoolChangerTrigger { timeout: 1s, task: None }, poisoned: false, .. }), channel_factory: None, template_to_job_id: TemplateToJobId { template_id_to_job_id: CircularBuffer { buffer: [], capacity: 10 }, request_id_to_template_id: {} }, req_ids: Id { state: 0 } }, poisoned: false, .. }), task_collector: Mutex(Mutex { data: [AbortHandle { id: Id(12) }, AbortHandle { id: Id(15) }], poisoned: false, .. }), coinbase_tx_prefix: Owned([]), coinbase_tx_suffix: Owned([]) }
STD ERR: thread 'tokio-runtime-worker' panicked at jd-client/src/lib/job_declarator/mod.rs:139:22:
STD ERR: LastDeclareJob not found
STD ERR: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
PanicInfo {
    payload: Any { .. },
    message: Some(
        explicit panic,
    ),
    location: Location {
        file: "src/external_commands.rs",
        line: 255,
        col: 21,
    },
    can_unwind: true,
    force_no_backtrace: false,
}
```

**Comment by plebhash** [comment]
hey @Shourya742 how about we do a recorded workshop next week to dive into this issue?

we have a workshop on youtube where I started working on #931: https://youtu.be/rNdm9Q4KTWE?si=90VJQ73Y-nlPvF-k

this would be part 2

**Comment by Shourya742** [comment]
> hey @Shourya742 how about we do a recorded workshop next week to dive into this issue?
> 
> we have [a workshop on youtube](https://youtu.be/rNdm9Q4KTWE?si=90VJQ73Y-nlPvF-k) where we started working on #931
> 
> this would be part 2

Yes, sure. Once I am done with the MG test for this scenario, we can hop on a call.

**Comment by plebhash** [comment]
@GitGab19 I want to revisit a discussion around one specific detail of this issue.

I was re-watching our [workshop recording](https://drive.google.com/file/d/1aNc4ZDrKw_rFnpx3C1DZUrRu67YSxlRq/view).

The description of issue #901 doesn't make it clear whether the bug is on JDC or the pool. As a consequence, during the workshop (around minute 12) we agreed that the MG test should have a real pool (as opposed to a mock).

However, PR #904 never touched anything on the pool implementation, which suggests the bug lives solely on JDC.

Now, I'm helping @Shourya742 write the MG test to reproduce this bug, and we are having to introduce extra complexity into the test, since we need to launch two separate TP mocks (one for real JDC and one for real pool).

So my question is: is it safe to reproduce the bug with a mock pool? That would remove the need for two separate TP mocks and greatly simplify the MG test.

**Comment by GitGab19** [comment]
Bug was on JDC side, Pool was just doing eveything right. 
So if it simplifies things, feel free to use a mocked pool for the MG test 👍

**Comment by plebhash** [comment]
> Bug was on JDC side, Pool was just doing eveything right. So if it simplifies things, feel free to use a mocked pool for the MG test 👍

that makes sense!

could you write a description of how this mock pool should behave? what messages should it expect, and what should it send back as a response?

**Comment by GitGab19** [comment]
Wait a second, shouldn't you just run the pool using our hosted TP in the config file? So that you don't have to mock a pool. Or I am missing something?

**Comment by plebhash** [comment]
setting this as blocked since @GitGab19 and I realized our mental model was wrong

we are doing some calls to refine it, and we will post a refined visual diagram of the message flow here once we are confident about it

**Comment by plebhash** [comment]
if this takes too much time, we should consider moving this to milestone 1.0.3

**Comment by plebhash** [comment]
@GitGab19 + @Shourya742 + @rrybarczyk just finished a call where we refined a diagram for the message flow of this issue

![#923 drawio](https://github.com/user-attachments/assets/b5310ebc-6ce6-4b3a-bbfb-e1f853178bf7)

hopefully this will help @Shourya742 unblock his work here

**Comment by pavlenex** [comment]
Should this one be dropped in favor of #1077 it seems there's already a WIP PR for it though @plebhash @Shourya742?

**Comment by Shourya742** [comment]
> Should this one be dropped in favor of #1077 it seems there's already a WIP PR for it though @plebhash @Shourya742?

@pavlenex , It makes sense to avoid writing MG tests for all these bug scenarios since integration tests will eventually cover them in a more reliable way. But getting the integration test suite to that level of maturity will take some time. We should keep these issues open, and as the test suite matures, we can add these scenarios to the test flows. Alternatively, we could take a test-driven development approach with the integration test suite—implementing features as needed to handle specific scenarios when they arise.
